ssg/oval_object_model/oval_document.py

```diff
@@ -0,0 +1,307 @@
+    )
+
+    product_name = generator_el.find("./{%s}product_name" % OVAL_NAMESPACES.get("oval"))
+    schema_version = generator_el.find(
+        "./{%s}schema_version" % OVAL_NAMESPACES.get("oval")
+    )
+    product_version = generator_el.find(
+        "./{%s}product_version" % OVAL_NAMESPACES.get("oval")
+    )
+    oval_document = OVALDocument(
+        product_name.text, schema_version.text, product_version.text
+    )
+
+    for definition_el in definitions_el:
+        oval_document.load_definition(definition_el)
+
+    for test_el in tests_el:
+        oval_document.load_test(test_el)
+
+    for object_el in objects_el:
+        oval_document.load_object(object_el)
+
+    if states_el:
+        for state_el in states_el:
+            oval_document.load_state(state_el)
+
+    if variables_el:
+        for variable_el in variables_el:
+            oval_document.load_variable(variable_el)
+
+    return oval_document
+
+
+class ExceptionDuplicateOVALEntity(Exception):
+    pass
+
+
+class OVALDocument(OVALBaseObject):
+    def __init__(self, product_name, schema_version, product_version):
+        self.oval_header_info = {
+            "product_name": product_name,
+            "schema_version": schema_version,
+            "product_version": product_version,
+        }
+
+        self.definitions: dict[str, Definition] = {}
+        self.tests: dict[str, Test] = {}
+        self.objects: dict[str, ObjectOVAL] = {}
+        self.states: dict[str, State] = {}
+        self.variables: dict[str, Variable] = {}
+
+    def _get_xml_element_from(self, shorthand):
+        valid_oval_xml_string = "{}{}{}".format(
+            oval_header, shorthand, oval_footer
+        ).encode("utf-8")
+        return ElementTree.fromstring(valid_oval_xml_string)
+
+    def load_shorthand(self, xml_string):
+        xml_element = self._get_xml_element_from(xml_string)
+
+        for child_node_el in xml_element.findall(
+            "./{%s}def-group/*" % OVAL_NAMESPACES.get("oval_def")
+        ):
+            if child_node_el.tag is ElementTree.Comment:
+                continue
+            elif child_node_el.tag.endswith("definition"):
+                self.load_definition(child_node_el)
+            elif child_node_el.tag.endswith("_test"):
+                self.load_test(child_node_el)
+            elif child_node_el.tag.endswith("_object"):
+                self.load_object(child_node_el)
+            elif child_node_el.tag.endswith("_state"):
+                self.load_state(child_node_el)
+            elif child_node_el.tag.endswith("_variable"):
+                self.load_variable(child_node_el)
+            else:
+                sys.stderr.write(
+                    "Warning: Unknown element '{}'\n".format(child_node_el.tag)
+                )
+
+    @staticmethod
+    def _is_external_variable(component):
+        return "external_variable" in component.tag
+
+    @staticmethod
+    def _add_oval_component(component, component_dict):
```

Comment:
This method feels too complex. Also, I'd prefer to build the exception messages using `.format()` instead of `+` operators.